Add `splitpath(p::AbstractString)` to Base.Filesystem.

NHDaly opened the pull request:
Opposite of `joinpath()`, it separates a path into its parts.

NHDaly committed:
Add `splitpath(p::AbstractString)` to Base.Filesystem.

Opposite of `joinpath()`, it separates a path into its parts.

StefanKarpinski commented on base/path.jl:
````diff
@@ -196,6 +197,29 @@ function pathsep(paths::AbstractString...)
+```
+"""
+function splitpath(p::AbstractString)
+    out = ()
````
This should return a `Vector{String}` not a tuple. That way you can also push stuff onto it.

NHDaly replied:
okay yeah, sounds good to me. i'll change that now! :)

NHDaly replied:
Done.

StefanKarpinski reviewed (commented):
I've definitely wanted this before, so 👍. Need to figure out what to do about drive letters on Windows 😭

NHDaly committed:
switch "is root" check to `ismount` && remove comments since code is self-explanatory

NHDaly commented:
I *think* that last push should fix the windows case, but i'd have to switch over to my bootcamp to really check (and/or add better windows-specific tests and wait for appveyor, but.. yeuckh)..

NHDaly committed:
Switch splitpath() to return an Array{String} instead of tuple

StefanKarpinski commented on base/path.jl:
```diff
@@ -211,13 +211,15 @@ julia> splitpath("/home/myuser/example.jl")
 """
 function splitpath(p::AbstractString)
-    out = ()
+    out = AbstractString[]
```
Better to make it just `String`, Postel's Law and all: this should take all kinds of strings but return one very specific kind of string vector, i.e. `Vector{String}`.

NHDaly replied:
>  I'm pretty sure `dirname` et al. always return `String` anyway.

```julia
help?> dirname
search: dirname

  dirname(path::AbstractString) -> AbstractString
```
That's actually the reason I made it an `AbstractString`, was to match `dirname`/`basename`.

... But actually those end up calling `splitdir`, which definitely does return a `String`. _And_, despite what the docstring says, it also only accepts a `String`?:
https://github.com/JuliaLang/julia/blob/a2fdb7d078f9281bbaf49133cf37f1344ab76b59/base/path.jl#L131-L137

NHDaly replied:
Well, I've changed my `splitpath` to match: Take and return `String`, but say `AbstractString` in the docs.. I want to match the style of the other functions.

If those should be changed, this can be done in the same batch. :)

quinnj commented:
Just leaving a reference here to HTTP.jl's [`splitpath`](https://github.com/JuliaWeb/HTTP.jl/blob/5630bf49e460a92b3b2fe49a5e5d51e7a3df5692/src/URIs.jl#L368) implementation; it's URI-specific, but I thought I'd mention it anyway.

NHDaly committed:
s/AbstractString/String/g

NHDaly committed:
return type simpliication

NHDaly commented on base/path.jl:
```diff
@@ -196,6 +197,31 @@ function pathsep(paths::AbstractString...)
+    out = String[]
+    while true
+        isempty(p) && break
+        ismount(p) && (push!(out, dirname(p)); break)
```
Ah, yeah, thanks! :)

NHDaly replied:
Done.

NHDaly committed:
realtimereview: reverse!、splitdir、[tests]

NHDaly commented on base/path.jl:
```diff
@@ -196,6 +197,31 @@ function pathsep(paths::AbstractString...)
+        push!(out, basename(p))
+        p = dirname(p)
+    end
+    return reverse(out)
```
Done.

NHDaly committed:
Attempt to add windows-focused splitpath unittests

StefanKarpinski commented on base/path.jl:
````diff
@@ -196,6 +197,32 @@ function pathsep(paths::AbstractString...)
+# Examples
+```jldoctest
+julia> splitpath("/home/myuser/example.jl")
+("/", "home", "myuser", "example.jl")
````
This output is going to be wrong now that this returns an array.

NHDaly replied:
👍 thanks! Fixed.

NHDaly committed:
Fix exports, which should make splitpath exported.

NHDaly commented on base/path.jl:
```diff
@@ -196,6 +197,32 @@ function pathsep(paths::AbstractString...)
+    while true
+        isempty(p) && break
+        d, b = splitdir(p)
+        ismount(p) && (push!(out, d); break)
```
Yeah you're definitely right about that. Good call. Actually, it looks like `dirname` already has the right drive logic for windows, so i think my original code was actually reasonable, `dirname(p) == p`. That is:
```
d == p && (push!(out, d); break)
```
That should be fine both for the unix and windows cases: dirname("/") returns "/" and dirname("C:") should return "C:".

Let me put it back to that and we can think more about it.

NHDaly committed:
replace `ismount` with `dirname(p) == p` check. Also cleanup.

Also restructured the while loop a bit to be more readable.

NHDaly commented:
> I think the right behavior for Windows is to split the drive off at the start and then concatenate it back with the first element in the path at the end.

Take another look when you can. I think the way it is now is actually best. It's generic for both unix and windows, and, I think, simple to understand!

NHDaly commented on test/path.jl:
```diff
@@ -85,6 +85,29 @@
+            @test ["/", "a", "b"] == splitpath("/a/b")
+            @test ["/"] == splitpath("/")
+            @test ["a"] == splitpath("a/")
+            @test ["a", "b"] == splitpath("a/b/")
```
Meep. I guess you're probably right:
```julia
julia> splitdir("a/")
>> ("a", "")
julia> splitdir("a///")
>> ("a", "")
```

I have never liked when utilities treat paths a with trailing slash differently, because tools seem to be inconsistent with when they add one. Haha oh, I guess I just described Postel's Law! Learned something new today. But yeah, like with Postel's Law, I think I wanted to be flexible with the input, so it would work whether or not there was a trailing slash. And I can't imagine why someone would want an empty path component in an array of the path components...

\*But\* I also think consistency across these functions is important, so I'm totally happy to go with whatever you think is best.

NHDaly replied:
Fwiw, python3 does not add the trailing `""`:

```python3
>>> from pathlib import PurePosixPath, PureWindowsPath
>>> PurePosixPath('/path/to/file.txt').parts
('/', 'path', 'to', 'file.txt')
>>>
>>> PurePosixPath('/path/to/dir/').parts
('/', 'path', 'to', 'dir')
```

StefanKarpinski replied:
Let's follow Python here—that's what I did for the rest of the path utils. I also think it's the right call, having thought about it some more: otherwise code that calls `splitpath` always has to look out for the trailing slash corner case. What was pushing me in the other direction was wanting to have some kind of identity that `joinpath(splitpath(path)) == path`, but I think avoiding the trailing slash gotcha is more important. That is probably why the Python function works the way it does.

NHDaly replied:
ah yeah that's too bad that `joinpath(splitpath(path))` can change the trailing slash....

I was going to say that actually `joinpath(splitpath(path)) == normpath(path)`, and it almost does, except for that it always trims trailing slashes. But I think that's okay, for all the reasons we just cited above. :) Thanks!

StefanKarpinski commented:
Yes, you're right—this is the way to go:
```jl
module Windows
    module Sys
        isunix() = false
        iswindows() = true
    end
    include("base/path.jl")
end

julia> Windows.splitpath(raw"C:\a\b\c")
4-element Array{String,1}:
 "C:\\"
 "a"
 "b"
 "c"
```
I was confused by the tests which are not quite right—they give the results you should get for relative paths for absolute paths on Windows. We should really make it possible to test the Windows and UNIX pure path operations on both kinds of systems—making sure the test are right is really annoying. The trick I have above lets you get a `Windows` module that has all the Windows versions of the path functions.

NHDaly committed:
Fixed windows tests: splitpath("C:\\a")->["C:\\", "a"]

NHDaly commented:
Oh cool, that's a neat trick! :)

And oh okay cool, thanks for pointing that out; I've fixed the windows tests to keep the slash in the drive name.

Also, through your test `Windows` module set up, I was able to see that windows handles unix-style "/" paths correctly as well, so I moved all those tests out of the `else` block and made them universal.

StefanKarpinski commented on test/path.jl:
```diff
@@ -89,22 +89,21 @@
-            @test ["J:"] == splitpath("J:\\\\")
+            @test ["C:\\", "a", "b", "c"] == splitpath("C:\\\\a\\b\\c")
+            @test ["C:\\"] == splitpath("C:\\\\")
+            @test ["J:\\"] == splitpath("J:\\\\")
```
Add a test for a relative path with a drive? (Yes, this is a thing, weirdly enough.)

NHDaly replied:
What do you mean by this? Can you give me an example?

StefanKarpinski replied:
`C:`, `C:a`, `C:a\b`

NHDaly replied:
Done, thanks!

StefanKarpinski commented:
Nice, this is looking really good! Aside from testing relative paths with drives on Windows, this seems pretty good to go to me. It might also be good to test a few more weird corner cases: drive-only, etc.

NHDaly committed:
Added more corner-case unit tests

fredrikekre commented on base/path.jl:
```diff
@@ -196,6 +197,32 @@ function pathsep(paths::AbstractString...)
+        dir, base = splitdir(p)
+        dir == p && (push!(out, dir); break)  # Reached root node.
+        if !isempty(base)  # Skip trailing '/' in basename
+            push!(out, base)
```
`pushfirst!` and skip the `reverse!` is clearer IMO (as suggested earlier).

NHDaly replied:
Done.

fredrikekre commented on base/path.jl:
```diff
@@ -196,6 +197,32 @@ function pathsep(paths::AbstractString...)
+        if !isempty(base)  # Skip trailing '/' in basename
+            push!(out, base)
+        end
+        p = dir;
```
Remove trailing `;`

NHDaly replied:
Done.

fredrikekre commented on base/precompile.jl:
```diff
@@ -122,6 +122,7 @@ precompile(Tuple{typeof(Base.Filesystem.isdir), String})
 precompile(Tuple{typeof(Base.Filesystem.pwd)})
 precompile(Tuple{typeof(Base.Filesystem.splitdir), String})
 precompile(Tuple{typeof(Base.Filesystem.splitext), String})
+precompile(Tuple{typeof(Base.Filesystem.splitpath), String})
```
I think this file is generated automatically so you can probably skip this line.

NHDaly replied:
Oh cool, okay thanks. Removed.

fredrikekre commented on test/path.jl:
```diff
@@ -85,6 +85,38 @@
     @test relpath(S(joinpath("foo","bar")), S("foo")) == "bar"
 
+    @testset "splitpath" begin
+        @test ["a", "b", "c"] == splitpath(joinpath("a","b","c"))
```
Ordering the test here and below as `@test splitpath(joinpath("a","b","c")) == ["a", "b", "c"]` is clearer IMO.

NHDaly replied:
Done, thanks! I should've noticed that the surrounding test style is the way you have it. Old habits from my last company. :) Thanks!

NHDaly committed:
Use pushfirst! not reverse!; switch @test order

NHDaly committed:
Added relative path C: drive tests

NHDaly commented:
PTAL

StefanKarpinski commented on test/path.jl:
```diff
@@ -85,6 +85,41 @@
 
+    @testset "splitpath" begin
+        @test splitpath(joinpath("a","b","c")) == ["a", "b", "c"]
+        @test splitpath("") == []
```
This seems wrong to me. In particular, note that `joinpath()` is an error, not `""`. This seems like it should be `[""]` instead of `[]`; `joinpath("") == ""` so you get the input back as you would want.

NHDaly replied:
hmm. Yeah I *think* that makes sense (from the identity viewpoint).

As another perspective, if I ask _"what are the directories and files in this here path: `""`?"_, then _"There aren't any! That is, it's these ones: `[]`"_ seems to me like the right response. Certainly it seems more plausible an answer than _"ah yes, it contains this file: `[""]`!"_.

Put another way, should `length(splitpath(""))` really be the same as `length(splitpath("a.txt"))`?

So I'm not sure what's best there. I guess it seems noteworthy that you can't get `splitpath` to return `[""]`, I just don't know if that's a good or a bad thing. I do agree that having a potential error lurking inside `joinpath(splitpath(p)...)` seems scary...

StefanKarpinski replied:
> There aren't any!

This isn't true though: `""` is the path of the current directory. It has a single component which happens to be the empty string, but that's not really relevant. You can replace the empty string in these cases  with `.` and everything should behave the same.

NHDaly replied:
Oh! Really? Yeah well, in that case i'm happy.

In that case, I think it does make sense that `length(splitpath(""))` is `1`. So `length(splitpath(p))` can never be `0`, and I think that makes sense. Okay yeah, I'll change this then. Thanks for explaining that!

NHDaly replied:
Done.

NHDaly commented on test/path.jl:
```diff
@@ -85,6 +85,41 @@
+            @test splitpath("J:\\") == ["J:\\"]
+            @test splitpath("C:") == ["C:"]
+            @test splitpath("C:a") == ["C:", "a"]
+            @test splitpath("C:a\\b") == ["C:", "a", "b"]
```
I agree with you, i think! I didn't really understand drives before.

Ah, okay, so the current behavior is a consequence of the way `splitdir` treats `"C:a"`:
```
julia> Windows.dirname("C:a")
"C:"
```
I'm now understanding that to mean "there is no directory above `a` in drive `C:`. The result of this function call is the empty string, but in drive `C:`." (That is, the _current directory_ in drive `C:`.) Is that right?

If so, then I agree, the result should be `C:a`.

To get that behavior, i think we can do what you originally recommended, which is to strip of the drive via `splitdrive` at the start, and then just tack it back on to the first element of the string at the end. Does that sound good to you?

StefanKarpinski replied:
Yes, I think that should work. Give it a try and see how the tests fare.

NHDaly replied:
Done. This worked great! :)

NHDaly committed:
Fix relative path w/ drive ("C:a"->["C:a"]), and empty path (""->[""]).

NHDaly commented:
Okay, so I've got this almost working. I made the changes we discussed above, and those test cases all work.

But now I'm baffled by two new unit test errors i've run across:
```julia
julia0.7> @test Windows.splitpath("C:\\\\a\\b\\c") == ["C:\\", "a", "b", "c"]
Test Failed at REPL[18]:1
  Expression: splitpath("C:\\\\a\\b\\c") == ["C:\\", "a", "b", "c"]
   Evaluated: ["C:\\\\a\\b\\", "c"] == ["C:\\", "a", "b", "c"]
ERROR: There was an error during testing
```
and the corresponding non-drive example:
```julia
julia0.7> Windows.splitpath("\\\\a\\b\\c")
>> 2-element Array{String,1}:
 "\\\\a\\b\\"
 "c"
```

Strangely, this seems to come from the behavior or `splitdir`:
```julia
julia0.7> Windows.splitdir("\\\\a\\b")
>> ("\\\\a\\b", "")
```

But it's specific to _two_ leading slashes (two _escaped_ slashes):
```julia
julia0.7> Windows.splitdir("\\a\\b")
>> ("\\a", "b")

julia0.7> Windows.splitdir("\\\\a\\b")
>> ("\\\\a\\b", "")

julia0.7> Windows.splitdir("\\\\\\a\\b")
>> ("\\\\\\a", "b")

julia0.7> Windows.splitdir("\\\\\\\\a\\b")
>> ("\\\\\\\\a", "b")
```

NHDaly commented:
Ah, okay, I've found the problem. The problem is that `Windows.splitdrive` somehow thinks the whole thing is a drive:
```julia
julia0.7> Windows.splitdrive("\\\\a\\b")
>> ("\\\\a\\b", "")
```

This looks like a bug to me.

StefanKarpinski commented:
@vtjnash, weren’t you complaining about these windows regexes recently?

NHDaly commented:
Huh, I don't know whether it's a bug or not actually..
This is the part of the regex that's matching:
`\\\\[^\\]+\\[^\\]+`
It's the second component, here:
https://github.com/JuliaLang/julia/blob/58754ddc592c18a08cf5a341ce7cc536b177d147/base/path.jl#L38

... is `\\a\b` somehow actually a valid drivename in windows?
If so, then the reason it's doing `splitpath("C:\\\\a\\b\\c")` wrong is because we're essentially calling `splitdrive` twice; once in `splitpath` and once for every call to `splitdir`. I'll probably want to change the implementation to not call `splitdrive` at the beginning. One way to fix it would be to call `splitdrive` manually on each element and compare it with the output of `splitdir`, and then put it back if the only thing `splitdir` is doing is popping off the drive?

Probably the Most Correct thing to do would be to make another version of `splitdir` that skips the `splitdrive` part, and to use that instead of `splitdir` in my implementation for `splitpath`. What do you think?

NHDaly commented:
:O woahhhhh

```python
import os
>>> os.path.dirname("\\a\\b")
'\\a'
>>> os.path.dirname("\\\\a\\b")
'\\\\a\\b'
```
_Weird_, man.
```python
>>> Path('\\\\a\\b\\c').parts
('\\\\a\\b\\', 'c')
>>> Path('C:\\\\a\\b\\c').parts
('C:\\', 'a', 'b', 'c')
```
Huh, so yeah, i guess we should respect the weirdness. So then yeah i think we'll have to change the way the function works, because right now, when we pull off the drivename at the beginning, it changes how the rest of the path parses. So i think probably the best change is to make a custom `splitdir_ignore_drive` type function that does the same thing as `splitdir` but without the `splitdrive` part. Does that sound good to you?

NHDaly committed:
Fix splitpath broken "C:\\\\a\\b" with `_splitdir_nodrive`.

NHDaly commented:
Okay, i pushed up a commit that does what i'm describing. I dunno about the naming for the added function, but this at least gives us something concrete to discuss.

All the tests should pass now! :)

StefanKarpinski commented:
Windows paths are _real_ weird.

NHDaly commented:
> Windows paths are _real_ weird.

Haha 😂 I know, right?

What do you think about the `_splitdir_nodrive` approach? I think it looks like the best option to me! (But i'm open to a better name for that helper function.)

(Also, the travis failure looks like a flake: ``ERROR: LoadError: `makedocs` encountered an error. Terminating build``)

vtjnash commented:
> Also, the travis failure

If it only happens on Travis Linux x64, then it's a doctest failure that's being caused here. You can examine those failures locally with `make docs doctest=true`)

> weren’t you complaining about these windows regexes recently?

Yes, some of those match tests are definitely in the wrong order (#27498). @omus had been looking into them more with me on Slack some time back.

StefanKarpinski commented:
Sorry, I haven't gotten a chance to look at this in detail; but I think the behavior is correct now and given that we can always refactor the implementation later.

NHDaly committed:
Fix docstring output to fix broken doctest

StefanKarpinski commented:
Bump, I don't think there's anything left to be done here, is there?

StefanKarpinski commented:
Oh yeah, I rebased this locally and fixed some merge conflicts and then pushed it.

StefanKarpinski commented:
Thanks for the great work on this, @NHDaly!

NHDaly commented:
Hooray! :D Thanks for the merge!

And thanks for your help and the back-and-forth! It was really fun essentially pair-programming on this to some extent! 😋
Cheers!

StefanKarpinski commented:
Would you be willing to make a PR adding a NEWS entry as well?

NHDaly commented:
> Would you be willing to make a PR adding a NEWS entry as well?

Ah, yeah, lemme take a stab at that!
https://github.com/JuliaLang/julia/pull/29448

PallHaraldsson commented:
1. I see in the tests: # Unix-style paths are understood by all systems.

I didn't really see from the code, how it passes those tests, but anyway, it that's really wanted (and they work on Windows, not just in Julia), do mixed paths also work, e.g. C:\path/to\something

If they do Julia code should also work with that (not sure it does currently).

2. I'm not sure it's wanted, maybe it's actively not wanted, but should splitpath, also normalize them?

StefanKarpinski commented:
You could always try it...
```jl
julia> Windows.splitpath("C:\\path/to\\something")
4-element Array{String,1}:
 "C:\\"
 "path"
 "to"
 "something"
```
The normalization question doesn't make sense if you think about for a bit.

NHDaly commented:
Haha to be fair, i didn't know about the cool `Windows.<function>` trick until you told me either!

But yep, they do, and also there are a few tests for it here:
https://github.com/JuliaLang/julia/pull/28156/files#diff-7629aea6e21fcaba2e27cc0a32eee2abR118

StefanKarpinski commented:
(The `Windows` trick only works if you define the module as I did above.)
